### ISSUE TITLE
 Onnx export failed.   SyncBatchNorm expected input tensor to be on GPU

### ISSUE THREAD

**Issue by Edwardmark** [ultralytics/yolov5#1018]
## ❔Question
I trained the model with SyncBatchNorm.
When I try to convert the model to onnx using the default configuration, I encountered the following error:

So I try to convert the model with gpu, I set device to 'cuda', but met another error:
ONNX export failure: All input tensors must be on the same device. Received cuda:0 and cpu  

## Additional context
So how to convert a model with syncbn to onnx?
Thanks in advance.

Best,
Edward.

**Comment by glenn-jocher** [comment]
@Edwardmark Hello, thank you for your interest in our work! This issue seems to lack the minimum requirements for a proper response, or is insufficiently detailed for us to help you. Please note that most technical problems are due to:

- **Your changes to the default repository**. If your issue is not reproducible in a new `git clone` version of this repository we can not debug it. Before going further run this code and ensure your issue persists:
```bash
sudo rm -rf yolov5  # remove existing
git clone https://github.com/ultralytics/yolov5 && cd yolov5 # clone latest
python detect.py  # verify detection
# CODE TO REPRODUCE YOUR ISSUE HERE
```

- **Your custom data.** If your issue is not reproducible with [COCO](https://cocodataset.org/#home) or [COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) data we can not debug it. Visit our [Custom Training Tutorial](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data) for guidelines on training your custom data. Examine `train_batch0.jpg` and `test_batch0.jpg` for a sanity check of training and testing data.

- **Your environment.** If your issue is not reproducible in one of the verified environments below we can not debug it. If you are running YOLOv5 locally, ensure your environment meets all of the [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies specified below.

If none of these apply to you, we suggest you close this issue and raise a new one using the **Bug Report template**, providing screenshots and **minimum viable code to reproduce your issue**. Thank you!


## Requirements

Python 3.8 or later with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) dependencies installed, including `torch>=1.6`. To install run:
```bash
$ pip install -r requirements.txt
```


## Environments

YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Google Colab Notebook** with free GPU: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a>
- **Kaggle Notebook** with free GPU: [https://www.kaggle.com/models/ultralytics/yolov5](https://www.kaggle.com/models/ultralytics/yolov5)
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) 
- **Docker Image** https://hub.docker.com/r/ultralytics/yolov5. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) ![Docker Pulls](https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker)

## Status

![CI CPU testing](https://github.com/ultralytics/yolov5/workflows/CI%20CPU%20testing/badge.svg)

If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are passing. These tests evaluate proper operation of basic YOLOv5 functionality, including training ([train.py](https://github.com/ultralytics/yolov5/blob/master/train.py)), testing ([test.py](https://github.com/ultralytics/yolov5/blob/master/test.py)), inference ([detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py)) and export ([export.py](https://github.com/ultralytics/yolov5/blob/master/models/export.py)) on MacOS, Windows, and Ubuntu.

**Comment by Edwardmark** [comment]
@glenn-jocher any clue on how to solve that?

**Comment by Edwardmark** [comment]
I just want to know how to export model trainded with syncBN. In export.py, we have to load the model to cpu, but it raises the error: 
```
raise ValueError('SyncBatchNorm expected input tensor to be on GPU')
```

**Comment by glenn-jocher** [comment]
@Edwardmark you might try replacing these with batchnorm2d layers in the 'Update model' section of export.py:
https://github.com/ultralytics/yolov5/blob/b7007d03b4264b3bb805211a568d5e30cbb605dc/models/export.py#L43-L47

**Comment by Edwardmark** [comment]
@glenn-jocher Thanks. I change the code as follows, is that correct：
```
if isinstance(m, nn.SyncBatchNorm):
            m.bn = nn.BatchNorm2d()  # assign bn
```

**Comment by glenn-jocher** [comment]
@Edwardmark seems right? How did it go?

**Comment by glenn-jocher** [comment]
@Edwardmark BTW can you submit a PR if this works well for you?

**Comment by Edwardmark** [comment]
@glenn-jocher It does not work, for we have to assign bn  parameters to BatchNorm2d, I have not figured it out yet.

**Comment by pekinghk** [comment]
> @glenn-jocher It does not work, for we have to assign bn parameters to BatchNorm2d, I have not figured it out yet.

Have you solved it? I meet the same error!

**Comment by pekinghk** [comment]
> > @glenn-jocher It does not work, for we have to assign bn parameters to BatchNorm2d, I have not figured it out yet.
> 
> Have you solved it? I meet the same error!

I have solved the problem.
You should assign all bn parameters to BatchNorm2d.
```
def convert_batchnorm(module, process_group=None):
    module_output = module
    if isinstance(module, torch.nn.SyncBatchNorm):
        module_output = torch.nn.BatchNorm2d(num_features=module.num_features,
                                                eps=module.eps, 
                                                momentum=module.momentum,
                                                affine=module.affine,
                                                track_running_stats=module.track_running_stats)
        if module.affine:
            with torch.no_grad():
                module_output.weight = module.weight
                module_output.bias = module.bias
        module_output.running_mean = module.running_mean
        module_output.running_var = module.running_var
        module_output.num_batches_tracked = module.num_batches_tracked
    for name, child in module.named_children():
        module_output.add_module(name, convert_batchnorm(child, process_group))
    del module
    return module_output
```

**Comment by glenn-jocher** [comment]
@pekinghk I'm a little confused, I think the model.fuse() operation that occurs before export should remove all batch norm layers from the entire model. Can you provide a minimum reproducible example of the commands you used to cause the ONNX export error?

**Comment by pekinghk** [comment]
> @pekinghk I'm a little confused, I think the model.fuse() operation that occurs before export should remove all batch norm layers from the entire model. Can you provide a minimum reproducible example of the commands you used to cause the ONNX export error?

Thanks for your answering. I don't konw why the model.fuse() operation doesn't work for me. In fact, I do my job 
based on https://github.com/WongKinYiu/ScaledYOLOv4 which forked your previous code. Maybe it's not your problem.

**Comment by glenn-jocher** [comment]
@pekinghk oh ok. Yes it's possible this has been resolved already in current master. If the problem originates in ScaledYOLOv4 you should probably raise an issue directly on that repo.

**Comment by 42kun** [comment]
thanks gods what a nice code!

**Comment by aichase** [comment]
> ```
> def convert_batchnorm(module, process_group=None):
>     module_output = module
>     if isinstance(module, torch.nn.SyncBatchNorm):
>         module_output = torch.nn.BatchNorm2d(num_features=module.num_features,
>                                                 eps=module.eps, 
>                                                 momentum=module.momentum,
>                                                 affine=module.affine,
>                                                 track_running_stats=module.track_running_stats)
>         if module.affine:
>             with torch.no_grad():
>                 module_output.weight = module.weight
>                 module_output.bias = module.bias
>         module_output.running_mean = module.running_mean
>         module_output.running_var = module.running_var
>         module_output.num_batches_tracked = module.num_batches_tracked
>     for name, child in module.named_children():
>         module_output.add_module(name, convert_batchnorm(child, process_group))
>     del module
>     return module_output
> ```

@pekinghk this code works for jit.trace exporting in cpu mode, but it changes the output of the jit model, don't know why yet.
However it works for onnx exporting, the output of onnx model is unchanged.
